torch/csrc/jit/tensorexpr/codegen.cpp

```diff
@@ -83,21 +83,136 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+c10::optional<size_t> bufSize(BufPtr buf) {
+  size_t size = elementSize(buf->dtype().scalar_type()) * buf->dtype().lanes();
+  for (auto& d : buf->dims()) {
+    if (!d->isConstant()) {
+      return c10::nullopt;
+    }
+    size = size * (*intValue(d));
+  }
+  return size;
+}
+
+// This algorithm takes the list of intermediate buffers and their liveness
+// ragnes, and returns the allocations of these buffers. A buffer 'A' can be
+// allocated in the memory (appears as a pair of 'A's in the allocation results)
+// or reuse another buffer such as 'B' (appears as ('A', 'B')). Specifically, we
+// linearly scan the intermediate buffers by the time they appear, and try to
+// assign it an existing non-occupied memory allocation. If there are no such
+// allocations available, we'll create memory for it. Once we are beyond the
+// liveness range of this buffer, we'll mark its corresponding memory allocation
+// as "up for grabs" for future reuse.
+std::vector<std::pair<BufPtr, BufPtr>> linearScan(
+    const std::unordered_set<BufPtr>& bufs,
+    const std::unordered_map<BufPtr, std::tuple<int32_t, int32_t>>&
+        buf_ranges) {
+  // Sort buffers by the time they appear.
+  std::vector<BufPtr> bufs_sorted(bufs.begin(), bufs.end());
+  auto sorting_function_by_start_time = [&buf_ranges](
+                                            BufPtr b1, BufPtr b2) -> bool {
+    return std::get<0>(buf_ranges.at(b1)) < std::get<0>(buf_ranges.at(b2));
+  };
+  std::sort(
+      bufs_sorted.begin(), bufs_sorted.end(), sorting_function_by_start_time);
+
+  // Map intermediate buffers to the most recently used memory if any.
+  std::list<BufPtr> mem_up_for_grabs;
+  std::unordered_map<BufPtr, BufPtr> buf_mem_map;
+  std::vector<std::pair<BufPtr, BufPtr>> buf_allocs;
+
+  auto sorting_function_by_end_time = [&buf_ranges](
+                                          BufPtr b1, BufPtr b2) -> bool {
+    return std::get<1>(buf_ranges.at(b1)) < std::get<1>(buf_ranges.at(b2));
+  };
+  std::vector<BufPtr> buf_to_release;
+  for (auto buf : bufs_sorted) {
+    auto start = std::get<0>(buf_ranges.at(buf));
+    auto end = std::get<1>(buf_ranges.at(buf));
+
+    // Release memory for buffers whose liveness range ends before the creation
+    // time of this buf.
+    // TODO: optimize in-place opererations and copy operations
+    buf_to_release.clear();
+    for (auto& mapped : buf_mem_map) {
+      auto buf_mapped = mapped.first;
+      auto end_buf_mapped = std::get<1>(buf_ranges.at(buf_mapped));
+      if (end_buf_mapped < start) {
+        buf_to_release.push_back(buf_mapped);
+      }
+    }
+
+    // Sort the buffers in the order of used time so the head of the release
+    // list contains the most recently used buf.
+    std::sort(
+        buf_to_release.begin(),
+        buf_to_release.end(),
+        sorting_function_by_end_time);
+    for (auto& buf_rl : buf_to_release) {
+      mem_up_for_grabs.push_front(buf_mem_map[buf_rl]);
```

Comment:
Nit: use `buf_mem_map.at(buf_rl)` instead of `buf_mem_map[buf_rl]`. The difference is that the latter one would insert a null element if it was absent in the map whereas the former one would raise an error.